packages/react-router-dom/__tests__/link-href-test.tsx

```diff
@@ -58,7 +58,7 @@ describe("<Link> href", () => {
           <MemoryRouter initialEntries={["/inbox/messages"]}>
             <Routes>
               <Route path="inbox">
-                <Route path="messages" element={<Link to="../../about" />} />
+                <Route path="messages" element={<Link to="../../../about" />} />
```

Comment:
isn't this link going completely out of the routes? `..` traverses the route hierarchy, so:

- the hierarchy at `/ `[inbox, messages]`
- `..` goes to `[inbox]`
- `../..` goes to `[]`
- `../../..` goes completely out of everything

can you help me understand?

Reply:
From the description, I though that was the point of this test — showing that even if you accidentally put too many `..` you’ll still get `/` instead of an error or something nonsensical.

Reply:
yeah that was my understanding as well from the test description, proving that links are "safe" insofar that you can't goof yourself by doing too many

Reply:
Looking at this test and the one above in the original commit they were added, it seems clear this is just about whether there are more `..` than parent `Route` components. The first test shows that `..` takes you to the parent `Route`, and the second shows that you can resolve `../../` to `/` even when there is no explicit `<Route path="/">`. So it does seem like `../../about` is enough to satisfy the goal of the test.

https://github.com/remix-run/react-router/blob/15cdd0da4a446e3746f65a05626969789fd94715/packages/react-router-dom/__tests__/link-href-test.tsx#L31-L58

Reply:
Interestingly, if you compare to the two tests that _did_ have `../../..` in the original commit, the extra `..` doesn't really make sense. If I change them to `../../about` they still pass. So maybe those are the ones that should be changed?

https://github.com/remix-run/react-router/blob/15cdd0da4a446e3746f65a05626969789fd94715/packages/react-router-dom/__tests__/link-href-test.tsx#L319-L332

https://github.com/remix-run/react-router/blob/15cdd0da4a446e3746f65a05626969789fd94715/packages/react-router-dom/__tests__/link-href-test.tsx#L435-L452

Reply:
I'm fine if we change them back 👍 .  Michael's on vacation this week so we won't be able to get any clarity on exactly what he may have envisioned as the purpose for the tests so probably best to leave them unchanged.  My main concern was that we weren't properly asserting the following line in `resolveTo`:

```js
    // If there are more ".." segments than parent routes, resolve relative to
    // the root / URL.
    from = routePathnameIndex >= 0 ? routePathnames[routePathnameIndex] : "/";
```

But - funny enough, even if we change that to `from = routePathnames[routePathnameIndex]` everything still works since `from` being `undefined` causes it to default to `"/"` in `resolvePath` anyway so we still get the same behavior.

As to the original commit intent - I think it makes sense that `../../about` still passes - I would argue that it's not asserting the _"more than"_ aspect of the description.  `../../about` should go back to `/about` because its correctly traversing the exact number of parents that it has.  Adding the third `..` tests the scenario that you use "more than".  I altered the test locally to show how I'm thinking about it:

```jsx
    test('<Link to=".."> with more .. segments than parent routes resolves to the root URL', () => {
      let renderer: TestRenderer.ReactTestRenderer;
      TestRenderer.act(() => {
        renderer = TestRenderer.create(
          <MemoryRouter initialEntries={["/inbox/messages"]}>
            <Routes>
              <Route path="inbox">
                <Route
                  path="messages"
                  element={
                    <>
                      <Link to="./about" />        // 1
                      <Link to="../about" />       // 2
                      <Link to="../../about" />    // 3
                      <Link to="../../../about" /> // 4
                    </>
                  }
                />
              </Route>
            </Routes>
          </MemoryRouter>
        );
      });

      expect(renderer.root.findAllByType("a").map((a) => a.props.href)).toEqual([
        "/inbox/messages/about", 
        "/inbox/about", 
        "/about",
        "/about",
      ]);
    });
```

* Link 2 and 3 assert the same thing - that we can traverse upwards with `..`
* Link 4 asserts something different - specifically when we have "_more than_"

Reply:
Ah sorry, I reviewed too quickly, the test says "more than parents" so I like this change.

Reply:
K, I’ll revert my revert.